File: apps/android-authenticator/protocol/src/main/kotlin/org/siwd/protocol/Encoding.kt
package org.siwd.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.time.Instant

val MAGIC: ByteArray = byteArrayOf(0x53, 0x49, 0x57, 0x44) // "SIWD"

/** Base58 alphabet used by Dash Platform identity IDs (Bitcoin-style). */
private const val BASE58_ALPHABET =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

fun utf8(s: String): ByteArray = s.toByteArray(StandardCharsets.UTF_8)

fun concat(vararg parts: ByteArray): ByteArray {
    val total = parts.sumOf { it.size }
    val out = ByteArray(total)
    var o = 0
    for (p in parts) {
        System.arraycopy(p, 0, out, o, p.size)
        o += p.size
    }
    return out
}

fun u8(n: Int): ByteArray {
    require(n in 0..0xff) { "u8 out of range: $n" }
    return byteArrayOf(n.toByte())
}

fun u16be(n: Int): ByteArray {
    require(n in 0..0xffff) { "u16 out of range: $n" }
    return byteArrayOf(((n ushr 8) and 0xff).toByte(), (n and 0xff).toByte())
}

fun u32be(n: Int): ByteArray {
    require(n >= 0) { "u32 out of range: $n" }
    return ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(n).array()
}

fun i64be(n: Long): ByteArray =
    ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN).putLong(n).array()

fun lengthPrefixedUtf8(s: String): ByteArray {
    val bytes = utf8(s)
    require(bytes.size <= 0xffff) { "UTF-8 field too long: ${bytes.size}" }
    return concat(u16be(bytes.size), bytes)
}

fun unixSecondsFromRfc3339(iso: String): Long {
    // Accept with or without fractional seconds
    val instant = Instant.parse(iso)
    return instant.epochSecond
}

fun decodeIdentityId(identityId: String): ByteArray {
    val raw = base58Decode(identityId)
    require(raw.size == 32) { "identity id must decode to 32 bytes, got ${raw.size}" }
    return raw
}

fun encodeIdentityId(raw: ByteArray): String {
    require(raw.size == 32) { "identity id raw must be 32 bytes" }
    return base58Encode(raw)
}

fun buildCanonicalBytes(input: CanonicalInput): ByteArray {
    val identityRaw = decodeIdentityId(input.identityId)
    val issued = unixSecondsFromRfc3339(input.issuedAt)
    val expires = unixSecondsFromRfc3339(input.expiresAt)
    return concat(
        MAGIC,
        u32be(PROTOCOL_VERSION),
        u8(input.network.wire),
        lengthPrefixedUtf8(input.origin),
        u8(input.action.wire),
        u8(input.bindingPolicy.wire),
        lengthPrefixedUtf8(input.requestId),
        input.nonce,
        i64be(issued),
        i64be(expires),
        identityRaw,
        lengthPrefixedUtf8(input.dpnsName),
        u32be(input.keyId),
    )
}

fun sha256(data: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-256").digest(data)

fun sha256d(data: ByteArray): ByteArray = sha256(sha256(data))

fun digestCanonical(input: CanonicalInput): ByteArray =
    sha256d(buildCanonicalBytes(input))

fun bytesToHex(bytes: ByteArray): String =
    bytes.joinToString("") { b -> "%02x".format(b.toInt() and 0xff) }

fun hexToBytes(hex: String): ByteArray {
    require(hex.length % 2 == 0) { "hex length must be even" }
    return ByteArray(hex.length / 2) { i ->
        hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
    }
}

fun base64UrlEncode(bytes: ByteArray): String {
    val b64 = java.util.Base64.getEncoder().withoutPadding().encodeToString(bytes)
    return b64.replace('+', '-').replace('/', '_')
}

fun base64UrlDecode(s: String): ByteArray {
    val padded = s.replace('-', '+').replace('_', '/')
    val pad = when (padded.length % 4) {
        0 -> ""
        2 -> "=="
        3 -> "="
        else -> error("invalid base64url length")
    }
    return java.util.Base64.getDecoder().decode(padded + pad)
}

/** Display form: strip trailing ".dash" parent label. */
fun displayDashName(full: String): String {
    val n = full.trim()
    return if (n.endsWith(".dash", ignoreCase = true)) {
        n.dropLast(".dash".length)
    } else {
        n
    }
}

private fun base58Encode(input: ByteArray): String {
    if (input.isEmpty()) return ""
    var zeros = 0
    while (zeros < input.size && input[zeros].toInt() == 0) zeros++
    val input58 = ByteArray(input.size * 2)
    var length = 0
    for (b in input) {
        var carry = b.toInt() and 0xff
        var j = 0
        var k = input58.size - 1
        while (k >= 0 && (carry != 0 || j < length)) {
            carry += 256 * (input58[k].toInt() and 0xff)
            input58[k] = (carry % 58).toByte()
            carry /= 58
            k--
            j++
        }
        length = j
    }
    var start = input58.size - length
    while (start < input58.size && input58[start].toInt() == 0) start++
    val sb = StringBuilder()
    repeat(zeros) { sb.append('1') }
    for (i in start until input58.size) {
        sb.append(BASE58_ALPHABET[input58[i].toInt()])
    }
    return sb.toString()
}

private fun base58Decode(input: String): ByteArray {
    if (input.isEmpty()) return ByteArray(0)
    val input58 = ByteArray(input.length)
    for (i in input.indices) {
        val c = input[i]
        val digit = BASE58_ALPHABET.indexOf(c)
        require(digit >= 0) { "invalid Base58 character: $c" }
        input58[i] = digit.toByte()
    }
    var zeros = 0
    while (zeros < input58.size && input58[zeros].toInt() == 0) zeros++
    val decoded = ByteArray(input.length)
    var length = 0
    for (b in input58) {
        var carry = b.toInt() and 0xff
        var j = 0
        var k = decoded.size - 1
        while (k >= 0 && (carry != 0 || j < length)) {
            carry += 58 * (decoded[k].toInt() and 0xff)
            decoded[k] = (carry % 256).toByte()
            carry /= 256
            k--
            j++
        }
        length = j
    }
    var start = decoded.size - length
    while (start < decoded.size && decoded[start].toInt() == 0) start++
    val out = ByteArray(zeros + (decoded.size - start))
    // leading zeros already 0
    System.arraycopy(decoded, start, out, zeros, decoded.size - start)
    return out
}

File: packages/protocol/src/encoding.ts
import bs58 from "bs58";
import { sha256 } from "@noble/hashes/sha2.js";
import type {
  Action,
  BindingPolicy,
  CanonicalInput,
  Network,
} from "./types.js";
import { PROTOCOL_VERSION } from "./types.js";

export const MAGIC = new Uint8Array([0x53, 0x49, 0x57, 0x44]); // "SIWD"

const NETWORK_BYTE: Record<Network, number> = {
  testnet: 0,
  mainnet: 1,
};

const ACTION_BYTE: Record<Action, number> = {
  register: 1,
  login: 2,
  link: 3,
};

const BINDING_BYTE: Record<BindingPolicy, number> = {
  identity_bound: 1,
  name_bound: 2,
};

export function concatBytes(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

export function u8(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xff) {
    throw new Error(`u8 out of range: ${n}`);
  }
  return new Uint8Array([n]);
}

export function u16be(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xffff) {
    throw new Error(`u16 out of range: ${n}`);
  }
  return new Uint8Array([(n >>> 8) & 0xff, n & 0xff]);
}

export function u32be(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xffffffff) {
    throw new Error(`u32 out of range: ${n}`);
  }
  return new Uint8Array([
    (n >>> 24) & 0xff,
    (n >>> 16) & 0xff,
    (n >>> 8) & 0xff,
    n & 0xff,
  ]);
}

/** Signed big-endian int64. */
export function i64be(n: bigint): Uint8Array {
  const buf = new ArrayBuffer(8);
  new DataView(buf).setBigInt64(0, n, false);
  return new Uint8Array(buf);
}

export function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

export function lengthPrefixedUtf8(s: string): Uint8Array {
  const bytes = utf8(s);
  if (bytes.length > 0xffff) {
    throw new Error(`UTF-8 field too long: ${bytes.length}`);
  }
  return concatBytes(u16be(bytes.length), bytes);
}

/**
 * Floor RFC 3339 / ISO-8601 timestamp to Unix seconds.
 * Rejects invalid dates.
 */
export function unixSecondsFromRfc3339(iso: string): bigint {
  const ms = Date.parse(iso);
  if (Number.isNaN(ms)) {
    throw new Error(`invalid timestamp: ${iso}`);
  }
  return BigInt(Math.floor(ms / 1000));
}

/**
 * Decode a Dash Platform identity id (Base58 of 32 raw bytes).
 * Accepts the Base58 alphabet used by current Platform SDKs (bs58).
 */
export function decodeIdentityId(identityId: string): Uint8Array {
  let raw: Uint8Array;
  try {
    raw = bs58.decode(identityId);
  } catch {
    throw new Error(`invalid identity id Base58: ${identityId}`);
  }
  if (raw.length !== 32) {
    throw new Error(
      `identity id must decode to 32 bytes, got ${raw.length}`,
    );
  }
  return raw;
}

export function encodeIdentityId(raw: Uint8Array): string {
  if (raw.length !== 32) {
    throw new Error(`identity id raw must be 32 bytes, got ${raw.length}`);
  }
  return bs58.encode(raw);
}

/**
 * Build canonical SIWD bytes per PROTOCOL.md §8.
 * JSON is never signed; only this layout is.
 */
export function buildCanonicalBytes(input: CanonicalInput): Uint8Array {
  if (input.nonce.length !== 32) {
    throw new Error(`nonce must be 32 bytes, got ${input.nonce.length}`);
  }
  if (!Number.isInteger(input.keyId) || input.keyId < 0 || input.keyId > 0xffffffff) {
    throw new Error(`invalid keyId: ${input.keyId}`);
  }

  const network = NETWORK_BYTE[input.network];
  if (network === undefined) {
    throw new Error(`unknown network: ${input.network}`);
  }
  const action = ACTION_BYTE[input.action];
  if (action === undefined) {
    throw new Error(`unknown action: ${input.action}`);
  }
  const binding = BINDING_BYTE[input.bindingPolicy];
  if (binding === undefined) {
    throw new Error(`unknown bindingPolicy: ${input.bindingPolicy}`);
  }

  const identityRaw = decodeIdentityId(input.identityId);
  const issued = unixSecondsFromRfc3339(input.issuedAt);
  const expires = unixSecondsFromRfc3339(input.expiresAt);

  return concatBytes(
    MAGIC,
    u32be(PROTOCOL_VERSION),
    u8(network),
    lengthPrefixedUtf8(input.origin),
    u8(action),
    u8(binding),
    lengthPrefixedUtf8(input.requestId),
    input.nonce,
    i64be(issued),
    i64be(expires),
    identityRaw,
    lengthPrefixedUtf8(input.dpnsName),
    u32be(input.keyId),
  );
}

/** digest = SHA-256(SHA-256(canonical_siwd_bytes)) */
export function sha256d(data: Uint8Array): Uint8Array {
  return sha256(sha256(data));
}

export function digestCanonical(input: CanonicalInput): Uint8Array {
  return sha256d(buildCanonicalBytes(input));
}

export function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new Error("hex length must be even");
  }
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

/** Unpadded base64url encode. */
export function base64urlEncode(bytes: Uint8Array): string {
  let bin = "";
  for (const b of bytes) {
    bin += String.fromCharCode(b);
  }
  return Buffer.from(bin, "binary")
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

/** Unpadded base64url decode. */
export function base64urlDecode(s: string): Uint8Array {
  const padded = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = padded.length % 4 === 0 ? "" : "=".repeat(4 - (padded.length % 4));
  const buf = Buffer.from(padded + pad, "base64");
  return new Uint8Array(buf);
}
